### ISSUE TITLE
Fix DB2 syntax for the profiler

### ISSUE THREAD

**Issue by TeddyCr** [open-metadata/OpenMetadata#6953]
DB2 profiler is throwing error when computing specific metrics

```
[2022-08-25 23:35:55] WARNING  {Profiler:core:253} - Error trying to compute column profile for col - (ibm_db_dbi.ProgrammingError) ibm_db_dbi::ProgrammingError: SQLNumResultCols failed: [IBM][CLI Driver][DB2/AIX64] SQL0440N  No authorized routine named “LEN” of type “FUNCTION” having compatible arguments was found.  SQLSTATE=42884 SQLCODE=-440
[SQL: /* {“app”: “OpenMetadata”, “version”: “0.11.5.0"} */
SELECT AVG(DOUBLE((LEN(col)))) AS mean, count(col) AS “valuesCount”, count(DISTINCT col) AS “distinctCount”, NULL AS anon_1, min(LEN(col)) AS “minLength”, NULL AS anon__1, max(LEN(col)) AS “maxLength”, sum(CASE WHEN (col IS NULL) THEN ? ELSE ? END) AS “nullCount”, NULL AS anon__2, NULL AS anon__3
FROM “schema“.table FETCH FIRST 1 ROWS ONLY]
[parameters: (1, 0)]
(Background on this error at: https://sqlalche.me/e/14/f405)
```

**Comment by upenbendre** [comment]
@TeddyCr here is our Slack chat regarding this issue for your reference:

[2022-08-25 23:35:49] WARNING  {Profiler:core:253} - Error trying to compute column profile for my_col2 - (ibm_db_dbi.ProgrammingError) ibm_db_dbi::ProgrammingError: SQLNumResultCols failed: [IBM][CLI Driver][DB2/AIX64] SQL0418N  The statement was not processed because the statement contains an invalid use of one of the following: an untyped parameter marker, the DEFAULT keyword, or a null value.  SQLSTATE=42610 SQLCODE=-418 [SQL: /* {“app”: “OpenMetadata”, “version”: “0.11.5.0"} */ SELECT AVG(DOUBLE((my_col2))) AS mean, count(my_col2) AS “valuesCount”, count(DISTINCT my_col2) AS “distinctCount”, min(my_col2) AS “min”, NULL AS anon_1, max(my_col2) AS “max”, NULL AS anon__1, sum(CASE WHEN (my_col2 IS NULL) THEN ? ELSE ? END) AS “nullCount”, STDDEV_POP(my_col2) AS stddev, sum(my_col2) AS “sum” FROM "schema ".my_table FETCH FIRST 1 ROWS ONLY] [parameters: (1, 0)] (Background on this error at: https://sqlalche.me/e/14/f405)
[12:04](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661452458867179)
[2022-08-25 23:35:55] WARNING  {Profiler:core:253} - Error trying to compute column profile for my_column - (ibm_db_dbi.ProgrammingError) ibm_db_dbi::ProgrammingError: SQLNumResultCols failed: [IBM][CLI Driver][DB2/AIX64] SQL0440N  No authorized routine named “LEN” of type “FUNCTION” having compatible arguments was found.  SQLSTATE=42884 SQLCODE=-440 [SQL: /* {“app”: “OpenMetadata”, “version”: “0.11.5.0"} */ SELECT AVG(DOUBLE((LEN(my_column)))) AS mean, count(my_column) AS “valuesCount”, count(DISTINCT my_column) AS “distinctCount”, NULL AS anon_1, min(LEN(my_column)) AS “minLength”, NULL AS anon__1, max(LEN(my_column)) AS “maxLength”, sum(CASE WHEN (my_column IS NULL) THEN ? ELSE ? END) AS “nullCount”, NULL AS anon__2, NULL AS anon__3 FROM "schema ".my_table FETCH FIRST 1 ROWS ONLY] [parameters: (1, 0)] (Background on this error at: https://sqlalche.me/e/14/f405)
[12:04](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661452499997099)
The DB2 profiler gave errors like these, it seems the profiler queries were not supported due to various reasons. e.g. the LEN function.
[12:05](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661452530455249)
They are coming from Profiler:core. Could you please tell me which program this is on your github repo? I am not able to find
￼
[Teddy](https://app.slack.com/team/U02MLBVPV2S)  [11:37 AM](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661494070588519)
Hey [@Upen Bendre](https://openmetadata.slack.com/team/U03TA35RA3V), hope you are well. This looks like a need on our end to adjust the syntax for DB2. I'll open a ticket and work on it. In the meantime you can find the metrics [here](https://github.com/open-metadata/OpenMetadata/tree/main/ingestion/src/metadata/orm_profiler/metrics)
￼GitHub
[OpenMetadata/ingestion/src/metadata/orm_profiler/metrics at main · open-metadata/OpenMetadata](https://github.com/open-metadata/OpenMetadata/tree/main/ingestion/src/metadata/orm_profiler/metrics)
Open Standard for Metadata. A Single place to Discover, Collaborate and Get your data right. - OpenMetadata/ingestion/src/metadata/orm_profiler/metrics at main · open-metadata/OpenMetadata (71 kB)
https://github.com/open-metadata/OpenMetadata/tree/main/ingestion/src/metadata/orm_profiler/metrics
￼
[Teddy](https://app.slack.com/team/U02MLBVPV2S)  [11:48 AM](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661494698932729)
Is it failing for all your tables or just this one specifically?
￼
[Upen Bendre](https://app.slack.com/team/U03TA35RA3V)  [11:49 AM](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661494796113699)
this was failing for all tables where that query was generated , so I made one change in /usr/local/lib/python3.9/site-packages/metadata/orm_profiler/orm/functions/length.py - to add the Db2 Dialect, that addressed SQLSTATE=42884 SQLCODE=-440.
[11:50](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661494818409019)
I have made a few more changes in similar fashion, which I will share in one go.
[11:50](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661494833409449)
For this one ‘SQLSTATE=42610 SQLCODE=-418’ , a query like this is generatedSELECT AVG(DOUBLE((LENGTH(segment)))) AS mean, count(segment) AS “valuesCount”, count(DISTINCT segment) AS “distinctCount”,    NULL AS anon_1, min(LENGTH(segment)) AS “minLength”,    NULL AS anon__1, max(LENGTH(segment)) AS “maxLength”,    sum(CASE WHEN (segment IS NULL) THEN ? ELSE ? END) AS “nullCount”,    NULL AS anon__2, NULL AS anon__3 FROM “DPS   “.my_table FETCH FIRST 1 ROWS ONLYNULL as anon__2 is not supported in DB2. So where the datatype is NULL I think this query should omit that entry, But I am still researching what the code fix should be.  metrics/static/count.py, profiler/default.py , metrics/static/null_count.py, profiler/default.py, orm/registry.py. Some of these programs don’t have a CASE clause for Dialects, so I am not sure if I should make a change that would impact all database types. (edited) 
￼
[Teddy](https://app.slack.com/team/U02MLBVPV2S)  [11:50 AM](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661494837344869)
sounds good, feel free to open a PR directly -- let me know if you need help there. (edited) 
[11:50](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661494847796619)
thank you very much ￼
[11:52](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661494920620689)
I've open [this issue](https://github.com/open-metadata/OpenMetadata/issues/6953) -- if you comment on it I can assign it to you. ￼
￼GitHub
[Fix DB2 syntax for the profiler · Issue #6953 · open-metadata/OpenMetadata](https://github.com/open-metadata/OpenMetadata/issues/6953)
DB2 profiler is throwing error when computing specific metrics [2022-08-25 23:35:49] WARNING {Profiler:core:253} - Error trying to compute column profile for colname - (ibm_db_dbi.ProgrammingErro... (75 kB)
https://github.com/open-metadata/OpenMetadata/issues/6953
￼1
￼
[Upen Bendre](https://app.slack.com/team/U03TA35RA3V)  [3:29 PM](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661507974999279)
Hi Teddy
￼
[Teddy](https://app.slack.com/team/U02MLBVPV2S)  [3:30 PM](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661508052100109)
hey [@Upen Bendre](https://openmetadata.slack.com/team/U03TA35RA3V), what's up ￼
￼
[Upen Bendre](https://app.slack.com/team/U03TA35RA3V)  [3:33 PM](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661508194739959)
[https://github.com/open-metadata/OpenMetadata/blob/04ede3b05b34bea9251a11aa2726c83[…]estion/src/metadata/orm_profiler/metrics/composed/null_ratio.py](https://github.com/open-metadata/OpenMetadata/blob/04ede3b05b34bea9251a11aa2726c83572aec002/ingestion/src/metadata/orm_profiler/metrics/composed/null_ratio.py)[https://github.com/open-metadata/OpenMetadata/blob/48f6553fb34dd8b22a2e9f1d59c73cc[…]ngestion/src/metadata/orm_profiler/metrics/static/null_count.py](https://github.com/open-metadata/OpenMetadata/blob/48f6553fb34dd8b22a2e9f1d59c73cc99228b8d8/ingestion/src/metadata/orm_profiler/metrics/static/null_count.py)And a few modules inside SQLAlchemy may be leading to the “Error trying to compute column profile for segment - (ibm_db_dbi.ProgrammingError) ibm_db_dbi::ProgrammingError: SQLNumResultCols failed: [IBM][CLI Driver][DB2/AIX64] SQL0418N The statement was not processed because the statement contains an invalid use of one of the following: an untyped parameter marker, the DEFAULT keyword, or a null value. SQLSTATE=42610 SQLCODE=-418”
￼GitHub
[OpenMetadata/null_ratio.py at 04ede3b05b34bea9251a11aa2726c83572aec002 · open-metadata/OpenMetadata](https://github.com/open-metadata/OpenMetadata/blob/04ede3b05b34bea9251a11aa2726c83572aec002/ingestion/src/metadata/orm_profiler/metrics/composed/null_ratio.py)
Open Standard for Metadata. A Single place to Discover, Collaborate and Get your data right. - OpenMetadata/null_ratio.py at 04ede3b05b34bea9251a11aa2726c83572aec002 · open-metadata/OpenMetadata (71 kB)
https://github.com/open-metadata/OpenMetadata/blob/04ede3b05b34bea9251a11aa2726c83572aec002/ingestion/src/metadata/orm_profiler/metrics/composed/null_ratio.py
￼GitHub
[OpenMetadata/null_count.py at 48f6553fb34dd8b22a2e9f1d59c73cc99228b8d8 · open-metadata/OpenMetadata](https://github.com/open-metadata/OpenMetadata/blob/48f6553fb34dd8b22a2e9f1d59c73cc99228b8d8/ingestion/src/metadata/orm_profiler/metrics/static/null_count.py)
Open Standard for Metadata. A Single place to Discover, Collaborate and Get your data right. - OpenMetadata/null_count.py at 48f6553fb34dd8b22a2e9f1d59c73cc99228b8d8 · open-metadata/OpenMetadata (71 kB)
https://github.com/open-metadata/OpenMetadata/blob/48f6553fb34dd8b22a2e9f1d59c73cc99228b8d8/ingestion/src/metadata/orm_profiler/metrics/static/null_count.py
[3:36](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661508393991739)
The particular table that I was trying to run profiler on, is empty. The query generated for metrics was :SELECT AVG(DOUBLE((LENGTH(segment)))) AS mean, count(segment) AS “valuesCount”, count(DISTINCT segment) AS “distinctCount”,    NULL AS anon_1, min(LENGTH(segment)) AS “minLength”,    NULL AS anon__1, max(LENGTH(segment)) AS “maxLength”,    sum(CASE WHEN (segment IS NULL) THEN ? ELSE ? END) AS “nullCount”,    NULL AS anon__2, NULL AS anon__3 FROM “DPS   “.my_table FETCH FIRST 1 ROWS ONLYThis is invalid SQL. I manually changed it to something like :SELECT AVG(DOUBLE((LENGTH(segment)))) AS mean, count(segment) AS “valuesCount”, count(DISTINCT segment) AS “distinctCount”,    min(LENGTH(segment)) AS “minLength”,    max(LENGTH(segment)) AS “maxLength”,    sum(CASE WHEN (segment IS NULL) THEN 1 ELSE 0 END) AS “nullCount” FROM “DPS   “.my_table FETCH FIRST 1 ROWS ONLY;  And it worked in my SQL Client (I am using DBeaver)
[3:37](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661508449575419)
I am going to post as many details of this as I can in the github issue you provided, but it looks like this may be a fix to SQL Alchemy
￼
[Teddy](https://app.slack.com/team/U02MLBVPV2S)  [3:38 PM](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661508526640309)
I see, the root cause of this error is an empty table?
[3:38](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661508538907019)
Thanks for the investigation BTW.
Only visible to you
￼
[Slackbot](https://app.slack.com/team/USLACKBOT)  [3:41 PM](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661508703000001) Pssst! I didn’t unfurl https://github.com/open-metadata/OpenMetadata/blob/48f6553fb34dd8b22a2e9f1d59c73cc99228b8d8/ingestion/src/metadata/orm_profiler/metrics/static/null_count.py because it was already shared in this channel quite recently (within the last hour) and I didn’t want to clutter things up.
Show Preview AnywayDo Nothing
￼
[Upen Bendre](https://app.slack.com/team/U03TA35RA3V)  [3:41 PM](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661508703160999)
I am not sure its because of an empty table yet. But I am sure that the way [https://github.com/open-metadata/OpenMetadata/blob/48f6553fb34dd8b22a2e9f1d59c73cc[…]ngestion/src/metadata/orm_profiler/metrics/static/null_count.py](https://github.com/open-metadata/OpenMetadata/blob/48f6553fb34dd8b22a2e9f1d59c73cc99228b8d8/ingestion/src/metadata/orm_profiler/metrics/static/null_count.py) line 47 works, that an empty table may have been one reason, perhaps what the SQLAlchemy function returns for an empty table, is not processed correctly by the orm_profiler function
[3:43](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661508802537769)
Also ‘SQLSTATE=42610 SQLCODE=-418’ is because of two issues in the SQL query, and both need a separate solution:
1.      CASE WHEN (segment IS NULL) THEN ? ELSE ? END - the ? here which needed to be a proper value.
2. NULL AS anon__1To get both of these, SQLAlchemy is used, but I think the issue is the way it is called by profiler
￼
[Teddy](https://app.slack.com/team/U02MLBVPV2S)  [3:56 PM](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661509612150279)
From looking around it might be due to DB2 needing statement in the query to be cast to the correct type of the table columns. So NULL AS anon_1 would need to match whatever type that column is in the table
￼
[Teddy](https://app.slack.com/team/U02MLBVPV2S)  [4:05 PM](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661510154936579)
Right now if a metric is not supported by a column data type (i.e. median for a VARCHAR  column) we will return None which result in NULL AS annon_n
[4:08](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661510334640999)
In [here](https://github.com/open-metadata/OpenMetadata/blob/2f3f60de0819ffdabc1d4738ce6700a84032f092/ingestion/src/metadata/orm_profiler/profiler/core.py#L245-L249) (if you can update the source code) could you try to replace that block with
            *[
                metric(column).fn()
                for metric in metrics
                if not metric.is_window_metric() and metric(column).fn() is not None
            ]
￼
[Teddy](https://app.slack.com/team/U02MLBVPV2S)  [4:16 PM](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661510804200609)
If the above fix works and you are going to make a PR for the above things have moved around a bit and this function is now [here](https://github.com/open-metadata/OpenMetadata/blob/87f8bcc4fa057a8bd04a6c930682fc97e64a2c85/ingestion/src/metadata/interfaces/sqa_interface.py#L397).Then for the length function error it should be a matter of adding the below in this [file](https://github.com/open-metadata/OpenMetadata/blob/main/ingestion/src/metadata/orm_profiler/orm/functions/length.py)
@compiles(LenFn, Dialects.Db2)
def _(element, compiler, **kw):
    """Handles length function for DB2"""
    return "LENGTH(%s)" % compiler.process(element.clauses, **kw)
￼
[Teddy](https://app.slack.com/team/U02MLBVPV2S)  [4:26 PM](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661511398118389)
BTW we have cool swag for merged PRs (Mug and Stickers) ￼
￼
[Upen Bendre](https://app.slack.com/team/U03TA35RA3V)  [4:29 PM](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661511545869149)
The LENGTH function fix already worked for me
￼1
￼
[Upen Bendre](https://app.slack.com/team/U03TA35RA3V)  [4:56 PM](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661513166799629)
“Error trying to compute column profile for my_column_name - name ‘column’ is not defined” - I am getting this now for all the columns. The exception on the same function
￼
[Teddy](https://app.slack.com/team/U02MLBVPV2S)  [4:57 PM](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661513257955809)
My bad it should be
            *[
                metric(col).fn()
                for metric in metrics
                if not metric.is_window_metric() and metric(col).fn() is not None
            ]
(edited)
[4:59](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661513358619799)
we changed col to column in 0.12
￼
[Upen Bendre](https://app.slack.com/team/U03TA35RA3V)  [5:03 PM](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661513594733249)
ok that error is gone, now the others errors which are due to how the generated SQL query ends up having ? instead of actual values:SELECT AVG(DOUBLE((LENGTH(column16)))) AS mean, count(column16) AS “valuesCount”, count(DISTINCT column16) AS “distinctCount”, min(LENGTH(column16)) AS “minLength”, max(LENGTH(column16)) AS “maxLength”, sum(CASE WHEN (column16 IS NULL) THEN ? ELSE ? END) AS “nullCount” FROM "schema ".my_table FETCH FIRST 1 ROWS ONLY]
[5:03](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661513619683459)
Error trying to compute column profile for column16 - (ibm_db_dbi.ProgrammingError) ibm_db_dbi::ProgrammingError: SQLNumResultCols failed: [IBM][CLI Driver][DB2/AIX64] SQL0418N  The statement was not processed because the statement contains an invalid use of one of the following: an untyped parameter marker, the DEFAULT keyword, or a null value.  SQLSTATE=42610 SQLCODE=-418
[5:04](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661513660158969)
the ? are actually the param_1 and param_2 coming from SQLAlchemy , but they don’t seem to get replaced with what I guess are numbers?
￼
[Upen Bendre](https://app.slack.com/team/U03TA35RA3V)  [5:09 PM](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661513966823149)
the changes you proposed, are they needed for run_window_metrics, run_table_metric, and similar changes to cover for “None” type in run_query_metrics and run_composed_metrics?
￼
[Teddy](https://app.slack.com/team/U02MLBVPV2S)  [5:12 PM](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661514142532859)
￼ most likely
￼
[Teddy](https://app.slack.com/team/U02MLBVPV2S)  [5:46 PM](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661516200306509)
For the ?  I think you need to do 2 things:
1.      in orm_profiler/orm/functions/random_num.py  [[here](https://github.com/open-metadata/OpenMetadata/blob/87f8bcc4fa057a8bd04a6c930682fc97e64a2c85/ingestion/src/metadata/orm_profiler/orm/functions/random_num.py)] you need to add the following
@compile(RandomNumFn, Dialects.Db2)
def _(*_, **__):
    return "ABS(RAND())*100"
2. in orm_profiler/orm/functions/modulo.py [[here](https://github.com/open-metadata/OpenMetadata/blob/87f8bcc4fa057a8bd04a6c930682fc97e64a2c85/ingestion/src/metadata/orm_profiler/orm/functions/modulo.py)] you need to add the following
[...]
@compiles(ModuloFn, Dialects.Db2)
def _(element, compiler, **kw):
    """Modulo function for specific dialect"""
    value, base = validate_and_compile(element, compiler, **kw)
    return f"MOD({value}, {base})"
(edited)
￼
[Upen Bendre](https://app.slack.com/team/U03TA35RA3V)  [6:23 PM](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661518404100279)
ok I am trying this. curious how does the modulo.py and random_num come in to play when its a metric query with a CASE expression.
￼
[Teddy](https://app.slack.com/team/U02MLBVPV2S)  [6:33 PM](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661518999968919)
it is used in sampling
￼
[Upen Bendre](https://app.slack.com/team/U03TA35RA3V)  [7:09 PM](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661521157624269)
I made the two changes and I commented all these lines:        self.run_static_metrics(col)         self.run_window_metrics(col)         self.run_query_metrics(col)         self.run_composed_metrics(col)I still got[2022-08-26 19:07:16] ERROR    {Profiler:core:282} - Error while running table metric for: my_table - (ibm_db_dbi.ProgrammingError) ibm_db_dbi::ProgrammingError: SQLNumResultCols failed: [IBM][CLI Driver][DB2/AIX64] SQL0418N  The statement was not processed because the statement contains an invalid use of one of the following: an untyped parameter marker, the DEFAULT keyword, or a null value.  SQLSTATE=42610 SQLCODE=-418 [SQL: /* {“app”: “OpenMetadata”, “version”: “0.11.5.0"} */ SELECT count(*) AS “rowCount”, ? AS “columnCount”, ? AS “columnNames” FROM "schema ".my_table FETCH FIRST 1 ROWS ONLY] [parameters: (39, ‘column1,column2,column3,column4,column5,column6,column7,column8,column9,column10,column11,co ... (188 characters truncated) ... column20,column21,column22,column23,column24,column25,column26,column27,column28,column29,column30’)] (Background on this error at: https://sqlalche.me/e/14/f405)
[7:09](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661521172188079)
Next I uncommented the call to self.run_static_metrics(col), despite the changes to random_num and modulo, I still got several of these messages:[2022-08-26 19:10:25] WARNING  {Profiler:core:253} - Error trying to compute column profile for column8 - (ibm_db_dbi.ProgrammingError) ibm_db_dbi::ProgrammingError: SQLNumResultCols failed: [IBM][CLI Driver][DB2/AIX64] SQL0418N  The statement was not processed because the statement contains an invalid use of one of the following: an untyped parameter marker, the DEFAULT keyword, or a null value.  SQLSTATE=42610 SQLCODE=-418 [SQL: /* {“app”: “OpenMetadata”, “version”: “0.11.5.0"} */ SELECT AVG(DOUBLE((LENGTH(column8)))) AS mean, count(column8) AS “valuesCount”, count(DISTINCT column8) AS “distinctCount”, min(LENGTH(column8)) AS “minLength”, max(LENGTH(column8)) AS “maxLength”, sum(CASE WHEN (column8 IS NULL) THEN ? ELSE ? END) AS “nullCount” FROM "schema ".my_table FETCH FIRST 1 ROWS ONLY] [parameters: (1, 0)][2022-08-26 19:10:26] WARNING  {Profiler:core:253} - Error trying to compute column profile for column13 - (ibm_db_dbi.ProgrammingError) ibm_db_dbi::ProgrammingError: SQLNumResultCols failed: [IBM][CLI Driver][DB2/AIX64] SQL0418N  The statement was not processed because the statement contains an invalid use of one of the following: an untyped parameter marker, the DEFAULT keyword, or a null value.  SQLSTATE=42610 SQLCODE=-418 [SQL: /* {“app”: “OpenMetadata”, “version”: “0.11.5.0"} */ SELECT AVG(DOUBLE((LENGTH(column13)))) AS mean, count(column13) AS “valuesCount”, count(DISTINCT column13) AS “distinctCount”, min(LENGTH(column13)) AS “minLength”, max(LENGTH(column13)) AS “maxLength”, sum(CASE WHEN (column13 IS NULL) THEN ? ELSE ? END) AS “nullCount” FROM "schema ".my_table FETCH FIRST 1 ROWS ONLY] [parameters: (1, 0)] (Background on this error at: https://sqlalche.me/e/14/f405) (edited) 
￼
[Teddy](https://app.slack.com/team/U02MLBVPV2S)  [7:39 PM](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661522982083849)
￼ we'll probably need to do more investigations. I've split both issues. This way you can make your PR for the LEN function and then we can continue investigating the other error
[7:39](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661522990628869)
Does it work for you?
￼
[Upen Bendre](https://app.slack.com/team/U03TA35RA3V)  [7:46 PM](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661523383461309)
I don’t mind if you keep a single issue. I just ran the profiler for one table, when I run it for the whole schema, I might run into more.
[7:47](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661523446770889)
btw for Db2 we also had to make the following changes:vi /usr/local/lib/python3.9/site-packages/metadata/orm_profiler/orm/registry.py Change line 53 to Db2 = “ibm_db_sa”vi /usr/local/lib/python3.9/site-packages/metadata/orm_profiler/orm/functions/conn_test.py ## Adding code for Db2 at line 42 @compiles(ConnTestFn, Dialects.Db2) def _(*_, **__):    return “SELECT 42 FROM SYSIBM.SYSDUMMY1"vi /usr/local/lib/python3.9/site-packages/metadata/orm_profiler/orm/functions/length.py # Add a dialect for Db2 where the LENGTH function syntax is used: @compiles(LenFn, Dialects.Db2)vi /usr/local/lib/python3.9/site-packages/ibm_db_sa/reflection.py Change line 236 to  return {‘text’: c.first()[0]} (edited) 
[7:48](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661523515352399)
vi /usr/local/lib/python3.9/site-packages/metadata/ingestion/source/database/db2.py (This one is debatable, but it helped us get past). ## Comment these lines:         if not isinstance(connection, Db2Connection):             raise InvalidSourceException(                 f”Expected Db2Connection, but got {connection}”             )
￼
[Teddy](https://app.slack.com/team/U02MLBVPV2S)  [7:49 PM](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661523560304269)
Yep, this one below we merged a PR yesterday
@compiles(ConnTestFn, Dialects.Db2)
def _(*_, **__):
   return "SELECT 42 FROM SYSIBM.SYSDUMMY1"
[7:49](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661523597977689)
BTW, what OS do you have?
￼
[Upen Bendre](https://app.slack.com/team/U03TA35RA3V)  [7:50 PM](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661523618627159)
Note that in /usr/local/lib/python3.9/site-packages/ibm_db_sa/reflection.py -> there is a get_table_comments(), but if you compare the ibm_db_sa version thats part of your openmetadata_ingestion package and the ibm_db_sa on pip, there is a difference. The get_table_comments() is not there is both.
[7:50](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661523657123859)
I am currently running the profiler from a Macbook and the openmeta server is on CentOS docker. But I have run into the same errors with the client on CentOS as well

New
￼
[Teddy](https://app.slack.com/team/U02MLBVPV2S)  [7:51 PM](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661523682954989)
mac M1?
[7:51](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661523687977609)
or intel?
￼
[Upen Bendre](https://app.slack.com/team/U03TA35RA3V)  [7:53 PM](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661523780393499)
Mac is Intel

￼
[Teddy](https://app.slack.com/team/U02MLBVPV2S)  [7:53 PM](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661523787299989)
thank you
[7:53](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661523817061229)
Let me do more digging into the error and get back to you if I have more questions
[7:54](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661523861945669)
Let me know if you are able to solve it. We'll try to ship fixes in 0.12.x so that you have a seemless experience with DB2 in OM. ￼ (edited) 
￼
[Upen Bendre](https://app.slack.com/team/U03TA35RA3V)  [8:52 PM](https://openmetadata.slack.com/archives/D040GJC6A2C/p1661527360405219)
thanks …

**Comment by upenbendre** [comment]
It would be good to have Ingestion and Profiling work for DB2 z/OS and DB2 LUW. ibm-db driver used by ibm_db_sa is supposed to work for both, as per my understanding.

**Comment by harshach** [comment]
@ayush-shah can you follow-up on this

**Comment by TeddyCr** [comment]
@harshach DB2 fixes are planned for 1.0. There were a few elements reported that are more involving than a simple syntax update 🙂
